### PR TITLE
feat(helm): Add repo alias support

### LINKED ISSUE #4681: Support helm chart placeholder repositories
**What would you like Renovate to be able to do?**

Support placeholders like this:
```yaml
dependencies:
  - name: generic-service
    version: 5.0.2
    repository: '@axoom-github'
    alias: app
```

Also support `alias:axoom-github` too as allowed in https://github.com/helm/helm/pull/2174

**Describe the solution you'd like**

It seems we need to check other `yaml` files within the repo for placeholder values. Is the location deterministic?

**Describe alternatives you've considered**

Currently we just skip placeholders, so they're not updated.

**Additional context**

Example source: https://github.com/AXOOM/Templates.WebService/blob/b7fca8e6a8aa94283a4aa4a1a98c92d77d6d4f55/content/helmfile.yaml#L6-L8

### PR DESCRIPTION
**Pull request by timja**
Add repo alias support

Closes https://github.com/renovatebot/renovate/issues/4681

Note: Only tested via unit tests, any docs need updating?

### CODE REVIEW

File: lib/manager/common.ts [hunk 1]
```diff
@@ -18,6 +18,7 @@ export interface ExtractConfig extends ManagerConfig {
   endpoint?: string;
   global?: any;
   gradle?: { timeout?: number };
+  helm?: { aliases?: Record<string, string> };
```
Comment (timja) [inline lib/manager/common.ts]: is this the right way to add additional config?
also there's this 
 https://github.com/renovatebot/renovate/blob/84d5632bfd8a8664bb62352300d5b530724eb00c/renovate-schema.json#L1057
but the key is `helm-requirements` which doesn't play well in javascript land, will it get auto converted to `helmRequirements` or am I on the wrong track?

**Review by adusumillipraveen — DISMISSED** [review]

**Comment by rarkins** [comment]
Thanks @timja. Can you give a description of how this is intended to work. e.g. are `aliases` needing to be configured manually in config, or can they be autodetected from the repo?

**Comment by timja** [comment]
> Thanks @timja. Can you give a description of how this is intended to work. e.g. are `aliases` needing to be configured manually in config, or can they be autodetected from the repo?

There's no way to detect them we could add the stable one by default, 
`stable: https://kubernetes-charts.storage.googleapis.com/` 

It would be expected to add something like:
```
{
  "helm": {
    "aliases": {
      "my-company": "https://my-registry.gcr.io/",
      "stable": "https://kubernetes-charts.storage.googleapis.com/"
    }
  }
}
```

**Comment by timja** [comment]
Also see this comment for more details:
https://github.com/renovatebot/renovate/issues/4681#issuecomment-543839808

**Comment by rarkins** [comment]
Isn't it possible to detect aliases from elsewhere in the repo? We can manage it even if it's in different files.

i.e. isn't "my-company" in your example configured somewhere else within your repo, instead of each developer needing to manually provision it locally on their computer?

BTW we can't support unstructured key:value JSON configuration like you've done because it breaks JSON-schema validation.

**Comment by timja** [comment]
> Isn't it possible to detect aliases from elsewhere in the repo? We can manage it even if it's in different files.

https://github.com/renovatebot/renovate/issues/4681#issuecomment-543041397
There's no standard unfortunately, each developer just does it once when they join and then it's there forever

> BTW we can't support unstructured key:value JSON configuration like you've done because it breaks JSON-schema validation.

could you link to a sample please?

**Comment by rarkins** [comment]
I like the top level option name you suggested - `aliases` - because that name could be reused for other purposes later if other package managers have similar requirements. The problem with making `aliases` an object with key/value pairs as you suggest is that we have no way to formally validate them with JSON schema. If we instead made it an array of objects, it could be like this:

```json
{
  "aliases": [
    {
      "aliasKey": "my-company",
      "aliasValue": "https://my-registry.gcr.io/"
    },
    {
      "aliasKey": "stable",
      "aliasValue": "https://kubernetes-charts.storage.googleapis.com/"
    }
  ]
}
```

This way our JSON schema can specify `aliases`, `aliasKey` and `aliasValue` and also that the latter two are to be strings. It's a bit less elegant though. 

A second challenge is we'd want to make aliases overridable, e.g. a default "stable" value built into Renovate but allowing people to override it in their config. With a native object like you suggest then overriding would happen automatically when you merge the objects whereas it would need to be done explicitly with the above JSON schema approach.

@ViceIce @JamieMagee @felixfbecker do any of you have opinions on this?

**Comment by rarkins** [comment]
@felixfbecker that's great advice, thanks. I found a reference for it here: https://json-schema.org/understanding-json-schema/reference/object.html

Based on this I think we may be able to keep @timja's proposed config structure.

**Comment by timja** [comment]
PTAL not sure if I've handled the schema side properly, but I tested the json schema with vs-code and it enforced the keyValue being a URL

**Comment by ViceIce** [comment]
@timja you have to update the schema again, pease run `yarn create-json-schema` and commit changes.

@rarkins I'm fine with the current alias definition.

File: bin/create-json-schema.js [hunk 2]
```diff
@@ -40,6 +40,9 @@ function createSingleConfig(option) {
   if (option.default !== undefined) {
     temp.default = option.default;
   }
+  if (option.additionalProperties !== undefined) {
+    temp.additionalProperties = option.additionalProperties;
```
Comment (rarkins) [inline bin/create-json-schema.js]: Great job adding this!

File: lib/config/definitions.ts [hunk 3]
```diff
@@ -525,6 +526,18 @@ const options: RenovateOptions[] = [
     mergeable: true,
     cli: false,
   },
+  {
+    name: 'aliases',
+    description: 'Aliases for registries, package manager specific',
+    type: 'object',
+    default: {
+      stable: 'https://kubernetes-charts.storage.googleapis.com/',
```
Comment (rarkins) [inline lib/config/definitions.ts]: I think let's move this default inside the Helm default config, not the global `aliases` default

File: lib/manager/helm-requirements/extract.ts [hunk 4]
```diff
@@ -50,6 +51,12 @@ export async function extractPackageFile(
     if (dep.repository) {
       res.registryUrls = [dep.repository];
       if (dep.repository.startsWith('@')) {
+        const alias = config.aliases[dep.repository.slice(1)];
```
Comment (rarkins) [inline lib/manager/helm-requirements/extract.ts]: Can we either add a documentation line describing the `slice(1)` or perhaps split it into two lines with a meaningful variable name if that's better?

File: lib/config/definitions.ts (hunk 3 above)
Comment (timja) [reply lib/config/definitions.ts]: How to do that?

Comment (rarkins) [reply lib/config/definitions.ts]: Add an `aliases` array inside the `helm-requirements` default config object in this file, and remove this `default`.

**Review by rarkins — CHANGES_REQUESTED** [review]
Move aliases default to inside helm-requirements

File: lib/manager/common.ts (hunk 1 above)
Comment (rarkins) [reply lib/manager/common.ts]: You can surround it in `'helm-requirements'` anywhere necessary

File: lib/config/definitions.ts (hunk 3 above)
Comment (rarkins) [reply lib/config/definitions.ts]: I've moved this now but you'll need to clean up some things locally (e.g. the json schema regeneration)

**Comment by rarkins** [comment]
Thanks @timja !

**Pull request merged**